### ISSUE TITLE
naming scheme for modules

### ISSUE THREAD

**Issue by cos** [kframework/k-legacy#1933]
I think there is some confusion regarding the naming scheme for modules. I will take `IMP` as an example.

In K < 4.0, the syntax was in `IMP-SYNTAX` and the semantics in `IMP`. Thus, `IMP` imported `IMP-SYNTAX`. `IMP-SYNTAX` was used directly to generate program parsers via some Java hacks which made sure that this is possible, e.g., restricting `Id` etc.

In K >= 4.0, we have more fine-grained modules so various parsers are generated from a specialized modules. So now we have the original syntax of the language, a module for parsing programs, a module for parsing rules, etc. The problem is that it seems to me we have messed up the naming of these modules a bit. We now have:

```
module IMP-COMMON
  contains the syntax of the language
endmodule

module IMP-SYNTAX 
  // despite its name, it is actually the module used for generating the program parser and contains
  // no syntactic productions, but only two imports like the ones below
  // this module could easily be auto-generated for most cases, unless the user
  // wants to do anything very funky for their program parser
  import IMP-COMMON
  imports DOMAINS-SYNTAX
endmodule

module IMP
  import IMP-COMMON
  contains the semantics of the language
endmodule
```

From what I see, the above does not follow any of the naming schemes discussed so far (e.g., the one here: https://github.com/kframework/k/wiki/!!!-Module-System). We ended with this naming scheme by equating correct "the `-SYNTAX` module **is used for generating** the program parser" with the incorrect "the `-SYNTAX` module **is** the program parser".

We might play with naming in the long run but, for now, could we make a simple change: let's name the module containing the entire syntax of the language `-SYNTAX` and the module for the program parser `-PROGRAMS`? We'll also need to propagate the naming change to `domain.k` which also has the same issue.
So:

```
 X-COMMON => X-SYNTAX
 X-SYNTAX => X-PROGRAMS
```

I know we are on a deadline and there are bigger issues than naming, but we're propagating this throughout the tutorial and we'll confuse students next semester. E.g., @omarz1, who is much more familiar than them with everything, posted this on gitter when looking at `lambda.k`:

> I looked at the lambda.k definition. I don't understand the way you're doing things now. You used to have a main module and a syntax module. Now you have the syntax in a commons module, while the syntax module just imports basic modules or something.
> no. syntax module imports common module
> but why did you choose to make it like that?

Again, it's not a big issue, so unless we have a quick consensus, I would drop this -- we don't really have time for long discussion.

@radumereuta, @bmmoore, @grosu?

@grosu, the current situation is clearly rather far from what we discussed and you described on the wiki, but at least some things will get cleaned up along with introducing module-qualified sorts and labels. @radumereuta has been working on adapting the front-end.

**Comment by omarzd** [comment]
@cos If confusing students next semester was the most urgent problem, then they should use the 3.6 release with its tutorials. Like you said, you don't have time for this.

**Comment by omarzd** [comment]
If I have a say in this, I like the suggestion by @cos that is:

```
 X-COMMON => X-SYNTAX
 X-SYNTAX => X-PROGRAMS
```

And `X-PROGRAMS` must be auto-generated unless overriden, just like the `super()` method  in a Java inheriting class.

**Comment by grosu** [comment]
@omarz1 , the problem with allowing students to use K3.6 is that they will have no support from us if they do so.  We are not going to add improvements to 3.6 anymore, not even fix bugs, simply because we do not have the resources to do it.  And I can tell you from experience that students WILL find lots of issues with K during the class.  That's why we must release K 4.0 before the semester starts.

**Comment by grosu** [comment]
@cos Many thanks for your thoughts on this.  I do not like the current notations either.  What is worse is that, again, it looks like the particular names of modules have meaning in the tool.  The same happened before in previous versions of K and our users and ourselves hated it, because this only leads to headaches and inconsistencies.  We should _stop_ doing this.  Actually, I thought that we agree on that, so I am quite surprised to see that again.

I think we need to discuss this in person.  Can you meet over hangouts (today?) to nail this down for good?  Anyway, here are my thoughts:

1) We should _never_ ever have the particular module names have special meaning in K!  For example, one should _not_ see any differences in any contexts when one calls the modules IMP-SYNTAX and IMP, vs IMP-MYSYNTAX and IMP-SEMANTICS, vs when one has an IMP-COMMON module or not, etc.  How one modularizes ones definition and what names the modules will have is simply one's own business, and K should not care at all about that.

2) I said it many times, and I say it again now: $PGM (and all configuration variables) should simply be a string, and one should explicitly use #parse to tell K which module to use for parsing it.  For example:

```
  configuration <T color="yellow">
                  <k color="green"> #parse($PGM, IMP-SYNTAX, Pgm) </k>
                  <state color="red"> .Map </state>
                </T>
```

3) The above being said, I think we should have some default conventions for beginners, so that they are not need to be aware of the complex K parsing machinery from the beginning.  For example, if they only say `$PGM:Pgm` in the configuration, then we should parse $PGM with a default module that K builds automatically for them.  That module should include all the syntax available in the current module except for syntax that we should exclude using attributes.  In the case of ID, for example, we already have the attribute `notInRules` (we may want to change its name, though).  Such attributes tell K how to build the default module that it will use for parsing programs.  That default module should have an automatically generated name that no user should ever know a priori or should ever see unless they explicitly request to see the generated KORE.  For example, something like `IMP-FOR-PARSING-PROGRAMS-#7534354`.

We can discuss these details when we meet, @cos, these are only some thoughts to start the discussion.  Let me know when you can discuss this.  I know you are busy with PLDI reviews.

**Comment by grosu** [comment]
@cos and I had a discussion about this this morning and we reached a compromise that seems to give us the best we can get with minimal changes.

1) We keep the definitions in the tutorial unchanged (vs K3.6) in what regards the modules.  In the case of IMP, for example, we still have a module called "IMP-SYNTAX" imported by a module "IMP", and the latter defines a configuration whose `<k/>` cell contents is `<k> $PGM:Pgm </k>`.  So from a user perspective, no change.

2) Under the hood, we are going to automatically generate several modules associated to each module (one for parsing programs, one for parsing configurations with symbolic stuff, one for parsing rules, etc.) for now.  In the future we may (and we will) change this, for example using module composition operations between the user-defined module and modules defined in kore.k or ekore.k.  But right now we generate such modules we need directly, and their specific name matters.  Due to the strong time limitations, let's not change this.  We agreed that -PROGRAMS is not necessarily the best name for the module used for parsing programs, because some of the inputs in some K definitions may not even be programs.  So let's call it -PARSING.

In other words, if you want to add new syntax for parsing programs to a module M that is not included in the syntax for parsing the semantics, you define an explicit module M-PARSING; the contents that you define in that module will be added to the additional automatically generated contents.

3) We are going to propagate this change within the domains.k module, but we are _not_ going to advertise the above.  That's because we want to buy ourselves time to design the entire module system properly and fully after the transition to K 4.0.

**Comment by cos** [comment]
solved by #1949
